xip/xipether.c

```diff
@@ -0,0 +1,449 @@
+		req.n.nlmsg_type = RTM_DELROUTE;
+	}
+
+	req.r.rtm_family = AF_XIA;
+	req.r.rtm_table = XRTABLE_LOCAL_INDEX;
+	req.r.rtm_protocol = RTPROT_BOOT;
+	req.r.rtm_type = RTN_LOCAL;
+	req.r.rtm_scope = RT_SCOPE_HOST;
+
+	req.r.rtm_dst_len = sizeof(*dst);
+	addattr_l(&req.n, sizeof(req), RTA_DST, dst, sizeof(*dst));
+
+	if (rtnl_talk(&rth, &req.n, 0, 0, NULL, NULL, NULL) < 0)
+		exit(2);
+	return 0;
+}
+
+static int do_local(int argc, char **argv, int to_add)
+{
+	struct xia_xid dst;
+	const char *dev;
+	char strid[2 * XIA_XID_MAX];
+	unsigned char lladdr[MAX_ADDR_LEN];
+	unsigned oif, addrlen;
+
+	if (argc != 2) {
+		fprintf(stderr, "Wrong number of parameters\n");
+		return usage();
+	}
+
+	if (strcmp(argv[0], "dev")) {
+		fprintf(stderr, "Wrong parameters\n");
+		return usage();
+	}
+
+	dev = argv[1];
+	oif = ll_name_to_index(dev);
+	if (!oif) {
+		fprintf(stderr, "Cannot find device '%s'\n", dev);
+		return -1;
+	}
+	addrlen = ll_index_to_addr(oif, lladdr, sizeof(lladdr));
+	if (!addrlen) {
+		fprintf(stderr, "Cannot find device address '%s'\n", dev);
+		return -1;
+	}
+	if (form_ether_xid(oif, lladdr, XIA_XID_MAX * 2, strid) < 0) {
+		fprintf(stderr, "Cannot form ether-xid from given parameters\n");
+		return -1;
+	}
+
+	xrt_get_ppal_id("ether", usage, &dst, strid);
+	return modify_local(&dst, to_add);
+}
+
+static int do_addlocal(int argc, char **argv)
+{
+	return do_local(argc, argv, 1);
+}
+
+static int do_dellocal(int argc, char **argv)
+{
+	return do_local(argc, argv, 0);
+}
+
+static int modify_neigh(struct xia_xid *dst, unsigned char *lladdr,
+	int lladdr_len, unsigned oif, int to_add)
+{
+	struct {
+		struct nlmsghdr 	n;
+		struct rtmsg 		r;
+		char   			buf[1024];
+	} req;
+
+	memset(&req, 0, sizeof(req));
+
+	req.n.nlmsg_len = NLMSG_LENGTH(sizeof(struct rtmsg));
+
+	if (to_add) {
+		/* XXX Does one really needs all these flags? */
+		req.n.nlmsg_flags = NLM_F_REQUEST|NLM_F_CREATE|NLM_F_EXCL;
+		req.n.nlmsg_type = RTM_NEWROUTE;
+		req.r.rtm_scope = RT_SCOPE_LINK;
+	} else {
+		req.n.nlmsg_flags = NLM_F_REQUEST;
+		req.n.nlmsg_type = RTM_DELROUTE;
+		req.r.rtm_scope = RT_SCOPE_NOWHERE;
+	}
+
+	req.r.rtm_family = AF_XIA;
+	req.r.rtm_table = XRTABLE_MAIN_INDEX;
+	req.r.rtm_protocol = RTPROT_BOOT;
+	req.r.rtm_type = RTN_UNICAST;
+
+	req.r.rtm_dst_len = sizeof(*dst);
+	addattr_l(&req.n, sizeof(req), RTA_DST, dst, sizeof(*dst));
+	addattr_l(&req.n, sizeof(req), RTA_LLADDR, lladdr, lladdr_len);
+	addattr32(&req.n, sizeof(req), RTA_OIF, oif);
+
+	if (rtnl_talk(&rth, &req.n, 0, 0, NULL, NULL, NULL) < 0)
+		exit(2);
+	return 0;
+}
+
+static int do_Xneigh_common(int argc, char **argv, int to_add)
+{
+	struct xia_xid dst;
+	char *str_lladdr;
+	char strid[XIA_MAX_STRID_SIZE];
+	unsigned char lladdr[MAX_ADDR_LEN];
+	int lladdr_len;
+	const char *dev;
+	unsigned oif;
+
+	if (argc != 4) {
+		fprintf(stderr, "Wrong number of parameters\n");
+		return usage();
+	}
+	if (strcmp(argv[2], "dev") || strcmp(argv[0], "lladdr")) {
```

Comment:
Flip these around to test for "lladdr" first, since that should be specified first.

Reply:
Note that it's not wrong the way you did it, it's just a style thing.